feat(rootfs.py): ot3 RootFS class

sotaoverride opened the pull request:
Class for controlling OT3 RootFS 

OT3 RootFS manipulation to manage updates, system restore etc. This would act as a helper class/script.

sotaoverride committed:
feat(rootfs.py): ot3 RootFS class

Class for controlling OT3 RootFS

sotaoverride committed:
feat(rootfs.py): add subparser

Add subparsers to this class

sotaoverride committed:
feat(rootfs.py): add a few more subparsers

Add a few more subparsers

sotaoverride committed:
fix(rootfs.py): fix formatting

Fix formatting, multiline comments

sotaoverride committed:
feat(setup.py): add simple setup file

Add simple setup.py

sotaoverride committed:
fix(get-pip.py): dont track get-pip.py

Stop tracking get-pip.py

sotaoverride committed:
fix: fix structure

Fix structure add __init__.py for package

sotaoverride committed:
fix(setup.py): add install requires

Add INSTALL_REQUIRES

sotaoverride committed:
fix(rootfs.py): fix comments

fix comments

sotaoverride committed:
fix(rootfs.py): rm redundant copy

rm redundant copy

sotaoverride committed:
RootFs formatting

sotaoverride committed:
feat(rootfs.py): add constants

Add constants for commandline args

sotaoverride committed:
fix(rootfs.py): fix params, remove debugging code

fixes

sfoster1 reviewed (changes requested):
While this utility script looks like it would work, the immediate next step is to make it part of the actual update server anyway. I think we should skip straight to that step, and have this script be in a new subpackage of the update server; call it `openembedded` and stand it next to `buildroot`. it doesn't have to have a lot in it, just a `__main__.py` that contains the script itself. We can then add the dependencies to pipenv.

In the future we'll still have to refactor some things - it would be nice to share some of the logic here - but this would be a better start.

sotaoverride committed:
feat(rootfs.py): move RootFs.py under otupdate

move RootFs.py under otupdate

sotaoverride committed:
feat(setup.py): add RootFS dependencies

Add RootFS util Dependencies

sotaoverride committed:
fix(rootfs.py): flake8 formatting

flake8 formatting

sotaoverride committed:
feat(__main__.py): add __main__.py and call RootFS from there

Add __main__.py and call RootFs from it

sotaoverride committed:
fix(__main__.py): fix comments

fix comments

sotaoverride committed:
fix(rootfs.py\): fix flake8 linint

Fix flake8 linitng

sotaoverride committed:
feat(rootfscli.py): refactor for adding pytests

Refactor to add pytests for CLI

sotaoverride committed:
feat(test_cli.py): add cli pytests

Add cli pytests for RootFS

sotaoverride committed:
feat(test_cli.py): fix lint

fix lint

Laura-Danielle commented on update-server/otupdate/openembedded/RootFS.py:
```diff
@@ -0,0 +1,91 @@
+
+@dataclass
+class RootFSInfo:
+    major: str
```
Can we make the major and minor versions integers?

Laura-Danielle commented on update-server/otupdate/openembedded/RootFS.py:
```diff
@@ -0,0 +1,91 @@
+
+class RootFS:
+    """ CONFIG vars, can be changed through the commandline arguments """
+    ROOTFS_PART1 = ''
```
Can we also make these configs into a dataclass?

Laura-Danielle replied:
And then you can initialize the dataclass with these default values when you call it in the class.

Laura-Danielle commented on update-server/otupdate/openembedded/RootFS.py:
```diff
@@ -0,0 +1,91 @@
+    ROOT_FS_PARTITION = ''
+    SD_CARD_MOUNT_POINT = '/media/mmcblk1p1'
+
+    def setPartition(self, arg, PartitionName):
```
Nit: typically we don't use camel case in python for functions. Here we would call it `set_partition` rather than `setPartition`

Laura-Danielle commented on update-server/otupdate/openembedded/RootFS.py:
```diff
@@ -0,0 +1,91 @@
+
+        subprocess.run(["fw_setenv", BOOT_SRC_CARVE_OUT, "boot="+PartitionName])
+
+    def getPartition(self):
```
Same nit: no camel case.

Laura-Danielle commented on update-server/otupdate/openembedded/RootFS.py:
```diff
@@ -0,0 +1,91 @@
+        else:
+            return None
+
+    def swapPartition(self, arg):
```
Same nit: no camel case.

Laura-Danielle commented on update-server/otupdate/openembedded/RootFS.py:
```diff
@@ -0,0 +1,91 @@
+            else:
+                RootFS.setPartition(arg, RootFS.ROOTFS_PART1)
+
+    def factoryRestore(self, arg):
```
Same nit: no camel case.

Laura-Danielle commented on update-server/otupdate/openembedded/RootFS.py:
```diff
@@ -0,0 +1,91 @@
+                        RootFS.SD_CARD_MOUNT_POINT+RootFS.BMAP_IMAGE, RootFS.DISK])
+
+    """ debug fuctions """
+    def printRootFSPartition(self, arg):
```
Same nit: no camel case.

Laura-Danielle reviewed (commented):
Just some suggestions

sotaoverride committed:
feat(test_bmap.py): add simple bmap write test

Add a simple bmap write test without writiting to an actual disk

sotaoverride committed:
feat(test_bmap.py): fix format for flake8

fix format for flake8

sotaoverride committed:
Merge branch 'ot3-rootfs-class' of github.com:Opentrons/opentrons into ot3-rootfs-class

sotaoverride committed:
fix(test_bmap.py): fix assert

Fix assert

sotaoverride committed:
feat(pipfile): add bmap-tools==3.6 to pipfile

add bmap-tools==3.6 to pipfile

sotaoverride committed:
typo

sotaoverride committed:
feat(pipfile.lock): update pipfile.lock

update pipfile.lock

sotaoverride committed:
fix(rootfs.py): remove camel case

Remove camel case

sotaoverride committed:
fix(test_bmap.py): fix function name

fix function name

sotaoverride committed:
fix(rootfs.py): fix self calls

Fix self calls

sotaoverride committed:
feat(test_bmap.py): add simple test using mock

Add simple test using mock

sotaoverride committed:
feat(rootfs.py): add inner data clas for RootFS config

Add innder data class for RootFS config

sotaoverride committed:
fix(test_bmap.py): fix for Lint

Fix for Lint, remove space

sotaoverride committed:
feat(test_bmap.py): add simple mock

Add simple mock to record method calls

sotaoverride committed:
fix(rootfs.py): fix typo

Fix typo

sotaoverride commented:
> While this utility script looks like it would work, the immediate next step is to make it part of the actual update server anyway. I think we should skip straight to that step, and have this script be in a new subpackage of the update server; call it `openembedded` and stand it next to `buildroot`. it doesn't have to have a lot in it, just a `__main__.py` that contains the script itself. We can then add the dependencies to pipenv.
> 
> In the future we'll still have to refactor some things - it would be nice to share some of the logic here - but this would be a better start.

the __main__.py in otupdate still needs a refactor, got the other bits mentioned above in

DerekMaggio commented on update-server/otupdate/openembedded/RootFS.py:
```diff
@@ -0,0 +1,100 @@
+        self.RootFSConfig_ = self.RootFSConfig('' , '', '' , '',
+                                               'mmcblk0', '', '', '/media/mmcblkp1')
+
+    def set_partition(self, arg, PartitionName):
```
Unless I am missing something it looks like `PartitionName` is a variable. 
If so, it should use snake case and be `partition_name`. 
Using the uppercase camel case denotes a class, so it looks like you are passing a class called `PartitionName` to the function

sfoster1 replied:
let's get some type hints in here also

DerekMaggio commented on update-server/otupdate/openembedded/RootFS.py:
```diff
@@ -0,0 +1,100 @@
+        out = subprocess.Popen(shlex.split("df /"),
+                               stdout=subprocess.PIPE).communicate()
+        m = re.search(r'(/[^\s]+)s', str(out))
+        if m:
```
This should be, `if match is not None`. If for some reason the search returned a `False` or `True` value or `1` or `0` this would get all messed up

DerekMaggio commented on update-server/otupdate/openembedded/RootFS.py:
```diff
@@ -0,0 +1,100 @@
+        """swap partitions get current partition
+           and swap it with the other available partition"""
+        currentPartition = self.get_partition()
+        if currentPartition is not None:
```
Should this log some kind of error if `currentPartition` is None?

DerekMaggio commented on update-server/otupdate/openembedded/RootFS.py:
```diff
@@ -0,0 +1,100 @@
+    def swap_partition(self, arg):
+        """swap partitions get current partition
+           and swap it with the other available partition"""
+        currentPartition = self.get_partition()
```
Coding Convention: You are swapping between snake and camel case

sotaoverride replied:
opps, though snake was just for function names. ill go snake all over.

DerekMaggio commented on update-server/otupdate/openembedded/RootFSCLI.py:
```diff
@@ -0,0 +1,50 @@
+        parser_debug.add_argument('--tt', '-t', type=str, nargs='?', help="test title",
+                                  default="OT3 RootFS Test",
+                                  const="OT# RootFS Test")
+        parser_debug.set_defaults(func=rfs.debug)
```
To confirm I am understanding correctly. This line is where you are mapping the cli command to the actually functionality that it is going to run?

sotaoverride replied:
yep, thats the hope here.

DerekMaggio commented on update-server/otupdate/openembedded/RootFSCLI.py:
```diff
@@ -0,0 +1,50 @@
+        subparsers = parser.add_subparsers()
+        # create Debug subcommand
+        parser_debug = subparsers.add_parser('debug', help='Debug')
+        parser_debug.add_argument('--tt', '-t', type=str, nargs='?', help="test title",
```
With this one if you call `-t` without a value, `t` will equal `OT# RootFS Tests`? If you call it with a value, `t` will equal `OT3 RootFS Test`?

sotaoverride replied:
yeah, this was just to give the tests a title.. title passed gets appended at the end a fixed string
print('ROOTFS_TEST_TITLE '+arg.tt+'\n')

Laura-Danielle commented on update-server/otupdate/openembedded/RootFS.py:
```diff
@@ -0,0 +1,100 @@
+        SD_CARD_MOUNT_POINT: str = '/media/mmcblk1p1'
+
+    def __init__(self):
+        self.RootFSConfig_ = self.RootFSConfig('' , '', '' , '',
```
When you have defaults set, you can just do: `RootFSConfig()`. It will pre-populate the date class with the defaults you made above.

Laura-Danielle replied:
Also, unless you _really_ want to make sure you can't access this dataclass you can just pull it out of `RootFS` and put it near `RootFSInfo`

sotaoverride replied:
cool, thanks. Ill start using those dataclass features. Ive kinda moved onto this branch now [openembedded-update-server ](https://github.com/Opentrons/opentrons/pull/8303), will get these changes in on there.

sfoster1 commented on update-server/otupdate/openembedded/RootFS.py:
```diff
@@ -0,0 +1,100 @@
+        subprocess.run(["fw_setenv", self.RootFSConfig_.BOOT_SRC_CARVE_OUT,
+                        "boot="+PartitionName])
+
+    def get_partition(self):
```
type hints for return

sfoster1 commented on update-server/otupdate/openembedded/RootFS.py:
```diff
@@ -0,0 +1,100 @@
+            ri = RootFSInfo(major, minor, mp)
+            return ri
+        else:
+            return None
```
consider `raise`ing here, this isn't really an acceptable failure case right

sfoster1 commented on update-server/otupdate/openembedded/RootFS.py:
```diff
@@ -0,0 +1,100 @@
+                        self.RootFSConfig_.DISK])
+
+    """ debug fuctions """
+    def print_rootFS_partition(self, arg):
```
I think it'll be a little more flexible if these are just returning strings and a caller prints it if they want; that way we can use this class going forward

sfoster1 commented on update-server/otupdate/openembedded/RootFS.py:
```diff
@@ -0,0 +1,100 @@
+        tmp = self.get_partition()
+        print('Current RootFS Partition '+tmp.disk+'\n')
+
+    def print_rootFS_config(self, arg):
```
same here - maybe return a string instead

sfoster1 commented on update-server/otupdate/openembedded/RootFSCLI.py:
```diff
@@ -0,0 +1,50 @@
+        # create Restore subcommand
+        parser_factoryRestore = subparsers.add_parser('restore', help='Restore')
+        parser_factoryRestore.add_argument('--wbm', '-b', nargs='?',
+                                           const='opentrons-ot3-image'
```
we have a surprisingly long line length (93 i think?) we should take advantage

sfoster1 commented on update-server/otupdate/openembedded/requirments.txt:
```diff
@@ -0,0 +1,10 @@
+bmap-tools==3.6
```
let's get rid of this since we're not using it

sotaoverride replied:
egg format for bmaptools is suppose to install it as a command line tool, and im hoping to use it some system restore functions. this pr use it in factory_restore function

sfoster1 replied:
sorry, i meant this entire file. it's not used anymore

sfoster1 commented on update-server/setup.py:
```diff
@@ -36,7 +36,19 @@ def get_version():
     "A server to update software and firmware on Opentrons robots")
 PACKAGES = find_packages(where='.', exclude=["tests.*", "tests"])
-INSTALL_REQUIRES = ['aiohttp==3.4.4']
+INSTALL_REQUIRES = [
```
transitive dependencies don't go in `INSTALL_REQUIRES`, just direct dependencies - when using `setup.py` it's up to `pip` to figure it out.

one thing we might want to think about is providing the openembedded part as an addon? i don't think it matters much though

sfoster1 commented on update-server/otupdate/openembedded/RootFS.py:
```diff
@@ -0,0 +1,100 @@
+        dev = os.stat('/')[stat.ST_DEV]
+        major = os.major(dev)
+        minor = os.minor(dev)
+        out = subprocess.Popen(shlex.split("df /"),
```
i'm pretty sure this will work but it feels like a kind of weird workaround; could we parse `/etc/fstab` or the result of `mount` or i'm sure there's something in `/proc` somewhere?

sotaoverride committed:
fix(rootfs.py): change to snake case all over and other small fixes

change to snake case plus other fixes

DerekMaggio commented on update-server/otupdate/openembedded/RootFS.py:
```diff
@@ -21,7 +21,7 @@ class RootFSInfo:
 class RootFS:
     @dataclass
-    class RootFSConfig:
+    class root_FS_config:
```
You are going to hate me. 
Classes should be capitalized camel case. For example, `AatirShouldHateDerek`
Variables, functions, and methods should be snake case. For example, `a_whole_lot`

sotaoverride committed:
fix(rootfs.py): fix case for pricate variable

Fix case for private vraible. Ran replace wrong last time around!!

sotaoverride committed:
fix(rootfs.py): snake case for variable name

Make variable name snake case

sotaoverride committed:
fix(rootfs.py): return string rather than print

Make dbug fucntions return strings rather than print

sotaoverride committed:
style(rootfscli.py): make line longer

Make code lines longer for an easier read.

sotaoverride committed:
refactor(rootfs.py): use mount instead of df

Use mount instead of df to find disk / is mounted on

sotaoverride committed:
fix(setup.py): remove transitive dependencies from setup.py

remove transitive dependencies from setup.py

sotaoverride committed:
refactor(rootfs.py): add type hints

add type hints

sotaoverride committed:
chore(rootfs.py): fix for lint

Fix fir lint

sotaoverride committed:
fix(rootfs.py): fix variable name!!!!!

fix variable name!!!!!!!

sotaoverride committed:
refactor(rootfs.py): fix varaible name, freakin replace all

Fix varaible names!!!

sfoster1 commented on update-server/Pipfile:
```diff
@@ -6,6 +6,7 @@ name = "pypi"
 [packages]
 "e1839a8" = {path = ".", editable = true}
 virtualenv = "*"
+bmap-tools = {editable = true, ref = "master", git = "https://github.com/intel/bmap-tools.git"}
```
given that we have this as a versioned requirement in `install_requires` in setup.py, can we make this a versioned requirement here as well?

sotaoverride replied:
think the issue with bmap-tools was that since it's not on pypi doing something like bmap-tools = xxxx wasn't working out. is that what you mean by a straight up version requirement here?

sfoster1 replied:
it was, but in that case how are we specifying it in the the `setup.py`?

sotaoverride committed:
fix(requirments.txt): remove requirments.txt

remove requirments.txt

sotaoverride committed:
fix(setup.py pipfile): remove bmaptools from update-server package

remove bmaptools from update-server package. utiltily made a part of opentrons image.

sotaoverride committed:
fix(pipfile): add ref as commit id for bmap-tools3.6

Cant find a way to add version for editable paths in pipfile, adding commit id for bmap-tools
version 3.6 to match version in setup.py

Laura-Danielle reviewed (commented):
Just one comment. Looks fine otherwise.

sotaoverride commented:
moving this to openembedded-update-server
